File: RB101/small_problems/easy_2/sum_or_product.rb
# Input: Ask user for integer larger than 0, then ask sum or product of all the numbers between 1 and that integer
# Output: A sentence with the sum or the product between 1 and the number provided

# Algo
# set a variable called number, include to input converted to integer
# Set a variable called operator, equal to input from asking a question explicitly (e.g. s for sum, p for product), downcase it just to be sure
# Create a loop for doing this, but replace the += or *= with the appropriate operator as it goes up (can't do this dynamically because + and * aren't somethign you can assign)
# I think creating two helper functions will be cleaner, and help separate the logic if I just have to type += or *= anyway

def sum(num)
  (1..num).inject(:+)
end

def product(num)
  (1..num).inject(:*)
end

user_num = 0
loop do
  puts "Please enter an integer larger than 0."
  user_num = gets.chomp.to_i

  # Since the to_i returns 0 if not an integer, and we specified higher than 0 anyway
  break if user_num > 0
  puts "That's not a number larger than 0."
end

operator = ''
loop do
  puts "Enter 's' to compute the sum, 'p' to compute the product. (From 1 to #{user_num}"
  operator = gets.chomp.downcase

  break if %w(s p).include?(operator)
  puts "That's not a valid choice."
end

if operator == "s"
  result = sum(user_num)
  operator = "sum"
elsif operator == "p"
  result = product(user_num)
  operator = "product"
else
  puts "You didn't pick 'p' or 's'"
end

puts "The #{operator} of the integers between 1 and #{user_num} is #{result}"

File: PY101/small_problems/easy_1/sum_or_product.py
# Write a program that asks the user to enter an integer greater than 0,
# then asks whether the user wants to determine the sum or the product of
# all numbers between 1 and the entered integer, inclusive.

# Example:
# Please enter an integer greater than 0: 5
# Enter "s" to compute the sum, or "p" to compute the product. s

# The sum of the integers between 1 and 5 is 15.

# ask for an integer, validate it's an integer above 0
# ask for the action you want to take
# create a result variable that is equal to 1
# iterate over the range inclusive of 1 to the number, either multiplying or
# summing as you go.

OPERATIONS = {
    's': "sum",
    'p': "product",
}

def validate_integer(num):
    while not type(num) == int:
        try:
            num = int(num)
        except ValueError:
            num = input('Please choose a valid integer: ')
    return num

def validate_operator(choice):
    while choice not in ['s', 'p']:
        choice = input('That is not a valid option.')
    return OPERATIONS.get(choice)

number = validate_integer(input('Please enter an integer greater than 0: '))

while number <= 0:
    number = validate_integer(input('Please enter an integer greater than 0: '))

operation = validate_operator(input('Enter "s" to compute the sum, or "p" to '
                                     'compute the product: '))

match operation:
    case 'sum':
        result = sum(range(1, number + 1))
    case 'product':
        result = 1
        for num in range(1, number + 1):
            result *= num
    case _:
        print('Something went horribly wrong')

print(f'The {operation} of the numbers between 1 and {number} is {result}')